service/dap/server.go

```diff
@@ -325,53 +353,114 @@ func (s *Server) handleRequest(request dap.Message) {
 	jsonmsg, _ := json.Marshal(request)
 	s.log.Debug("[<- from client]", string(jsonmsg))
 
+	if _, ok := request.(dap.RequestMessage); !ok {
+		s.sendInternalErrorResponse(request.GetSeq(), fmt.Sprintf("Unable to process non-request %#v\n", request))
+		return
+	}
+
+	// These requests, can be handeled regardless of whether the targret is running
 	switch request := request.(type) {
-	case *dap.InitializeRequest:
-		// Required
-		s.onInitializeRequest(request)
-	case *dap.LaunchRequest:
-		// Required
-		s.onLaunchRequest(request)
-	case *dap.AttachRequest:
-		// Required
-		s.onAttachRequest(request)
 	case *dap.DisconnectRequest:
 		// Required
 		s.onDisconnectRequest(request)
+		return
+	case *dap.PauseRequest:
+		// Required
+		// TODO: implement this request in V0
+		s.onPauseRequest(request)
+		return
 	case *dap.TerminateRequest:
 		// Optional (capability ‘supportsTerminateRequest‘)
 		// TODO: implement this request in V1
 		s.onTerminateRequest(request)
+		return
 	case *dap.RestartRequest:
 		// Optional (capability ‘supportsRestartRequest’)
 		// TODO: implement this request in V1
 		s.onRestartRequest(request)
-	case *dap.SetBreakpointsRequest:
-		// Required
-		s.onSetBreakpointsRequest(request)
-	case *dap.SetFunctionBreakpointsRequest:
-		// Optional (capability ‘supportsFunctionBreakpoints’)
-		// TODO: implement this request in V1
-		s.onSetFunctionBreakpointsRequest(request)
-	case *dap.SetExceptionBreakpointsRequest:
-		// Optional (capability ‘exceptionBreakpointFilters’)
-		s.onSetExceptionBreakpointsRequest(request)
+		return
+	}
+
+	s.runningMu.Lock()
+	defer s.runningMu.Unlock()
+
+	// Most requests cannot be processed while the debuggee is running.
+	// We have a couple of options for handling these without blocking
+	// the request loop indefinitely when we are in running state.
+	// --1-- Return a dummy response or an error right away.
+	// --2-- Halt execution, process the request, resume execution.
+	// TODO(polina): do this for setting breakpoints
+	// --3-- Handle such requests asynchronously and let them block until
+	// the process stops or terminates (e.g. using a channel and a single
+	// goroutine to preserve the order). This might not be appropriate
+	// for requests such as continue or step because they would skip
+	// the stop, resuming execution right away. Other requests
+	// might not be relevant anymore when the stop is finally reached, and
+	// state changed from the previous snapshot. The user might want to
```

Comment:
Thanks for the detailed explanation. The first part of the current version matches what I expected - processing the requests that can be served while the program is running. I wanted to avoid many goroutines fighting for the lock. But now it returns err for all requests (except threads and breakpoints in the future),  so there will be effectively only one or two extra goroutine outstanding in any given time and we don't have to worry about requests' ordering.